src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using NuGet.Common;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Model;
+
+namespace NuGet.Protocol.Resources
+{
+    internal class VulnerabilityInfoResourceV3 : IVulnerabilityInfoResource
+    {
+        private readonly SourceRepository _sourceRepository;
+
+        internal VulnerabilityInfoResourceV3(SourceRepository sourceRepository)
+        {
+            _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
+        }
+
+        internal async Task<IReadOnlyList<V3VulnerabilityIndexEntry>> GetVulnerabilityFilesAsync(SourceCacheContext cacheContext, ILogger log, CancellationToken cancellationToken)
+        {
+            Uri vulnerabilityIndexUrl = await GetIndexUrlAsync(cancellationToken);
+            HttpSourceResource httpSourceResource = await _sourceRepository.GetResourceAsync<HttpSourceResource>(cancellationToken);
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(vulnerabilityIndexUrl.OriginalString, "vuln_index", httpSourceCacheContext);
+            IReadOnlyList<V3VulnerabilityIndexEntry>? vulnFiles = await httpSourceResource.HttpSource.GetAsync(request,
+                result =>
+            {
+                using (var textReader = new StreamReader(result.Stream))
+                using (var jsonReader = new JsonTextReader(textReader))
+                {
+                    var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IReadOnlyList<V3VulnerabilityIndexEntry>>(jsonReader);
+                    return Task.FromResult(parsed);
+                }
+            },
+                log,
+                cancellationToken);
+
+            if (vulnFiles == null)
+            {
+                string message = string.Format("Package source '{0}' vulnerability data index '{1}' did not successfully deserialize. This indicates a problem with the JSON file contents returned from the server.",
+                    _sourceRepository.PackageSource.Name,
+                    vulnerabilityIndexUrl.OriginalString);
+                throw new FatalProtocolException(message);
+            }
+
+            return vulnFiles;
+
+            async Task<Uri> GetIndexUrlAsync(CancellationToken cancellationToken)
+            {
+                ServiceIndexResourceV3 serviceIndex = await _sourceRepository.GetResourceAsync<ServiceIndexResourceV3>(cancellationToken);
+                return serviceIndex.GetServiceEntryUri(ServiceTypes.VulnerabilityInfo);
```

Comment:
From https://github.com/NuGet/NuGet.Client/pull/5115#issuecomment-1491185896 since GitHub doesn't provide threaded/resolvable conversations unless linked to a specific line in a file:

> May be the decision is already made, have we considered an option to download package vulnerabilities just like package versions by appending package name to the registration base URL for example https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/nuget.common/index.json. I don't have the performance metrics yet, but I think if the vulnerabilities database size increases in NuGet.org then performance may degrade.

@kartheekp-ms the design spec considered different options, and also for different considerations/use-cases, like performance: https://github.com/NuGet/Home/blob/dev/proposed/2022/vulnerabilities-in-restore.md

Basically, even if the file(s) keep getting larger and larger over time, the latency penalty of making many requests far outweighs any reduced size benefit. Consider the performance of `dotnet list package --vulnerable --include-transitive`. Adding that much delay into restore is unacceptable. It needs to be much, much faster.  But also it wouldn't meet some other design goals outlined in the spec.